// stryker-jasmine-runner/src/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_runner_1 = require("stryker-api/test_runner");
var JasmineConstructor = require("jasmine");
function toStrykerTestResult(specResult, timeSpentMs) {
    var status = test_runner_1.TestStatus.Failed;
    var failureMessages;
    if (specResult.status === 'disabled' || specResult.status === 'pending' || specResult.status === 'excluded') {
        status = test_runner_1.TestStatus.Skipped;
    }
    else if (!specResult.failedExpectations || specResult.failedExpectations.length === 0) {
        status = test_runner_1.TestStatus.Success;
    }
    else {
        failureMessages = specResult.failedExpectations.map(function (failedExpectation) { return failedExpectation.message; });
    }
    return {
        name: specResult.fullName,
        status: status,
        failureMessages: failureMessages,
        timeSpentMs: timeSpentMs
    };
}
exports.toStrykerTestResult = toStrykerTestResult;
function evalGlobal(body) {
    var fn = new Function('require', body);
    fn(require);
}
exports.evalGlobal = evalGlobal;
exports.Jasmine = JasmineConstructor;
